### PR TITLE
#167484604 Search functionality

### PR DESCRIPTION
**Pull request by malaba6**
### What does this PR do?
- Adds search Functionality and Custom filtering based on parameters

### Description of Task to be completed?
- Create an article helper to handle filter functionality
- Add the function to filter query params according to user input
- Add tests to search with filter functionality
You can search articles with the following query parameters:
- title
- author
- tagList

### How should this be manually tested?
In Postman or Swagger or any other tool of your choice visit these endpoints:
*GET*`/api/v1/articles?title=XXXval&author=XXXval&tagList=XXXval` To filter fetching articles

### What are the relevant pivotal tracker stories?
[search Functionality and Custom filtering based on parameters](https://www.pivotaltracker.com/story/show/#167484604)

### Screenshots? 
*In Postman*
Screen Shot 2019-09-04 at 15 43 50

### CODE REVIEW

**Review by mkiterian — DISMISSED** [review]
I don't see functionality to filter by key words

File: src/helpers/articles/articleHelper.js [hunk 1]
```diff
@@ -30,4 +32,35 @@ const uploadImage = async (filename) => {
   }
 };
 
-export { slugGen, uploadImage };
+/**
+ * @description filters query parameters
+ * @param {string} title
+ * @param {string} author
+ * @param {object} tagList
+ * @returns {object} filtered wuery params
```
Comment (Jkadhuwa) [inline src/helpers/articles/articleHelper.js]: Typo `@returns` think you meant query.

**Review by Jkadhuwa — CHANGES_REQUESTED** [review]
When I provide tags or author not who has no articles in the DB it returns an empty array `articles: []`. Can you display a message if no articles are found?

**Comment by malaba6** [comment]
> When I provide tags or author not who has no articles in the DB it returns an empty array `articles: []`. Can you display a message if no articles are found?

Fixed!

File: src/helpers/articles/articleHelper.js (hunk 1 above)
Comment (malaba6) [reply src/helpers/articles/articleHelper.js]: > Typo `@returns` think you meant query.

Yeah right! Fixed...

**Review by placideirandora — CHANGES_REQUESTED** [review]
1. When a user searches something which does not exist, you should respond with a meaningful message that the thing cannot be found. For instance, I was trying to search an invalid article title, you should have responded that the invalid article (actual value) does not exist instead of returning an empty array. 
Screen Shot 2019-09-05 at 12 38 31

2. I was not able to retrieve the article even if I provided the right article title and author. Therefore, you should fix this.
Screen Shot 2019-09-05 at 12 45 09
Screen Shot 2019-09-05 at 12 45 21

3. I was able to retrieve the article by searching with invalid property and value. You should fix this as well.
Screen Shot 2019-09-05 at 12 48 15

**Comment by malaba6** [comment]
> 1. When a user searches something which does not exist, you should respond with a meaningful message that the thing cannot be found. For instance, I was trying to search an invalid article title, you should have responded that the invalid article (actual value) does not exist instead of returning an empty array.
> 
> Screen Shot 2019-09-05 at 12 38 31
> 
> 1. I was not able to retrieve the article even if I provided the right article title and author. Therefore, you should fix this.
> 
> Screen Shot 2019-09-05 at 12 45 09
> 
> Screen Shot 2019-09-05 at 12 45 21
> 
> 1. I was able to retrieve the article by searching with invalid property and value. You should fix this as well.
> 
> Screen Shot 2019-09-05 at 12 48 15

Change made...

**Pull request merged**